CV2-6146: Add a message/conversation_id placeholder

melsawy opened the pull request:
## Description

Create a placeholder variable like {{message_id}} that can be included in a tipline message.

- [x] Implement a method called replace_placeholders in smooch_messages concern.
- [x] Call that method in the send_message_to_user implementations we have for the Zendesk provider and for the WhatsApp Cloud API provider:
- [x] Implement automated tests for the model method and for the two cases above

References: CV2-6146

## How to test?

Implement unit test.

## Checklist

- [x] I have performed a self-review of my code and ensured that it is safe and runnable, that code coverage has not decreased, and that there are no new Code Climate issues. I have also followed [Meedan's internal coding guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/pages/1309605889/Coding+guidelines).

melsawy commented:
@caiosba can you please test against `CAPI` already I verified against `Zendesk` provider 
![Screenshot from 2025-03-24 14-32-50](https://github.com/user-attachments/assets/de79f2fb-b50a-43d2-afc7-f324cf7fa19d)

caiosba commented on app/models/concerns/smooch_capi.rb:
```diff
@@ -228,6 +228,7 @@ def capi_send_message_to_user(uid, text, extra = {}, _force = false, preview_url
       payload = {}
       account, to = uid.split(':')
       return if account != self.config['capi_phone_number']
+      text = self.replace_placeholders(uid, text)
```
@melsawy , it's better to move this line to inside the condition below, no?

caiosba commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -568,5 +568,10 @@ def min_number_of_words_for_tipline_long_text
     end
+
+    def replace_placeholders(uid, text)
+      external_id = TiplineMessage.where(uid: uid).last&.external_id
```
For performance reasons, I suggest you run this query only if the message contains the placeholder.

melsawy replied:
Done and also I apply changes to allow multiple placeholders so in future it'll be easy to add a new placeholders.

caiosba commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -568,5 +568,10 @@ def min_number_of_words_for_tipline_long_text
     end
+
+    def replace_placeholders(uid, text)
+      external_id = TiplineMessage.where(uid: uid).last&.external_id
```
Just to be super safe, I suggest you filter by `team_id` as well.

melsawy replied:
Done

caiosba commented on test/models/bot/smooch_7_test.rb:
```diff
@@ -660,4 +660,13 @@ def teardown
     end
   end
+
+  test "should replace message_id placeholder" do
+    uid = random_string
+    tm = create_tipline_message uid: uid
+    text = 'Foo {{message_id}} bar {{message_id}}'
+    external_id = tm.external_id
+    output = Bot::Smooch.replace_placeholders(uid, text)
+    assert_equal "Foo #{external_id} bar #{external_id}", output
+  end
```
For the test, I suggest you set an explicit `external_id` and assert it at the end, for example:

```ruby
tm = create_tipline_message uid: uid, external_id: 'abc123'
[...]
assert_equal "Foo abc123 bar abc123", output
```

melsawy replied:
Done

caiosba commented on app/models/concerns/smooch_messages.rb:
```diff
@@ -568,5 +568,10 @@ def min_number_of_words_for_tipline_long_text
     end
+
+    def replace_placeholders(uid, text)
+      external_id = TiplineMessage.where(uid: uid).last&.external_id
```
Another suggestion after testing locally: Please also filter by `direction: 'incoming'` and `state: 'received'`, this way we can consistently get "the last message sent by that user".

caiosba reviewed (changes requested):
Thanks Sawy, it works well locally for CAPI. I left some comments to change a few things, and I'm happy to re-review after that. Another thing I noticed it's missing: let's make sure to shorten URLs in the message if the placeholder is found.

melsawy commented:
@caiosba please re-test against CAPI and also I applied a new changes to allow multiple placeholders in future

caiosba reviewed (changes requested):
Sawy, please re-request review once the URL shortening case is fixed (and I can test it locally for CAPI as well).

melsawy commented:
> Thanks Sawy for addressing all code comments from the previous review! There is still one thing missing though: if URL shortening is enabled, the URL will be shortened before the placeholders are replaced, which in practice will "remove" the placeholder from the message. So, can you please also call the `replace_placeholders` method in the [URL shortening method?](https://github.com/meedan/check-api/blob/develop/lib/url_rewriter.rb#L25)
I could not call replace_placeholders in `shorten_and_utmize_urls` method as I depend on `uid` so I called it before calling `shorten_and_utmize_urls` method

caiosba commented:
> > Thanks Sawy for addressing all code comments from the previous review! There is still one thing missing though: if URL shortening is enabled, the URL will be shortened before the placeholders are replaced, which in practice will "remove" the placeholder from the message. So, can you please also call the `replace_placeholders` method in the [URL shortening method?](https://github.com/meedan/check-api/blob/develop/lib/url_rewriter.rb#L25)
> > I could not call replace_placeholders in `shorten_and_utmize_urls` method as I depend on `uid` so I called it before calling `shorten_and_utmize_urls` method

OK, that's actually better.

caiosba reviewed (approved):
@melsawy, I pushed a commit to fix it for messages that have menus and buttons. Works well for CAPI. Please test for Zendesk as well.

melsawy commented:
> @melsawy, I pushed a commit to fix it for messages that have menus and buttons. Works well for CAPI. Please test for Zendesk as well.

Works well for Zendesk thanks @caiosba
